File: merge-intervals/merge-intervals.py
from interval import *


# return a new output array consisting of the merged intervals
# intervals input is sorted by starting time
def merge_intervals(intervals):
    if not intervals:
        return None

    output = [intervals[0]]

    for curr_interval in intervals[1:]:
        last_interval = output[-1]

        if curr_interval.start <= last_interval.end:
            last_interval.end = max(last_interval.end, curr_interval.end)
        else:
            output.append(curr_interval)

    return output

File: merge-intervals/merge-intervals.js
import { Interval } from "./_utils/interval.js";

export function mergeIntervals(intervals) {
	let result = [intervals[0]];

	for (let i = 1; i < intervals.length; i++) {
		const currInterval = intervals[i];
		const lastMergedInterval = result[result.length - 1];

		if (currInterval.start <= lastMergedInterval.end) {
			lastMergedInterval.end = Math.max(
				lastMergedInterval.end,
				currInterval.end
			);
		} else {
			result.push(currInterval);
		}
	}

	return result;
}
